src/core/exec-invoke.c

```diff
@@ -2548,6 +2557,173 @@ static int create_many_symlinks(const char *root, const char *source, char **sym
         return 0;
 }
 
+static int set_exec_storage_quota(int fd, uint32_t proj_id, const QuotaLimit *ql) {
+        int r;
+        uint64_t block_limit = 0, inode_limit = 0;
+
+        assert(fd >= 0);
+        assert(ql);
+
+        if (ql->quota_absolute == 0 || ql->quota_scale == 0)
+                block_limit = 1, inode_limit = 1;
+        else if (ql->quota_absolute == UINT64_MAX) {
+                struct statvfs disk_st;
+                r = fstatvfs(fd, &disk_st);
+                if (r < 0)
+                        return -errno;
+
+                block_limit = (uint64_t) DIV_ROUND_UP((int)((double) (disk_st.f_frsize * disk_st.f_blocks) / UINT32_MAX * ql->quota_scale), QIF_DQBLKSIZE);
+                inode_limit = (uint64_t) ((double) disk_st.f_files / UINT32_MAX * ql->quota_scale);
+        } else
+                block_limit = (uint64_t) DIV_ROUND_UP(ql->quota_absolute, QIF_DQBLKSIZE);
+
+        struct dqblk req = {
+                .dqb_bhardlimit = block_limit,
+                .dqb_bsoftlimit = 0,
+                .dqb_ihardlimit = inode_limit,
+                .dqb_isoftlimit = 0,
+                .dqb_valid = QIF_LIMITS
```

Comment:
please always insert a trailing comma when doing multi-line compound struct initializations. we do that as a matter of style, to keep later patches small that add new fields.